### ISSUE TITLE
`Write-Debug` expands `-Message` when it should `SilentlyContinue`

### ISSUE THREAD

**Issue by iRon7** [PowerShell/PowerShell#20738]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

`Write-Debug` and `Write-Verbose` expand the `-Message` where the commands should `SilentlyContinue`.

**Use case:**
I am using a class with has a lazy evaluation (because not all results always require to be evaluated).
It concerns a recursive script where I would like to add some `Write-Debug` commands, like:

```PowerShell
Write-Debug "This takes 3 seconds $(Sleep 3)"
```

### Expected behavior

```console
When the -Debug switch is omitted for my script the script shouldn't take more time because of the debug commands (in other words, the `-Message` should not be evaluated).
```

### Actual behavior

```console
When the -Debug switch is omitted for my script, the `-Message` is evaluated and takes 3 secondes (for the example).
```

### Error details

```console
Workaround:

Function test {
    [CmdletBinding()] param() 
    if ($DebugPreference -in 'Stop', 'Continue', 'Inquire') { Write-Debug "Test $(Sleep 3)"
}
```
```


### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.3.10
PSEdition                      Core
GitCommitId                    7.3.10
OS                             Microsoft Windows 10.0.19045
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by rhubarb-geek-nz** [comment]
I expect the evaluation of the message is done before [Write-Debug](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/write-debug?view=powershell-7.4) is invoked.

For this to work you would need Write-Debug to do the expansion and formatting depending on the current environment.

However the parameter `-Message` is defined as a `string` so the parameter expansion is going to try and evaluate it to a single string before invoking the cmdlet.

If it was defined as an `object` and `Write-Debug` used `ToString` then you could have delayed the expansion by creating an object where the message is expanded and formatted in the ToString method.

**Comment by iRon7** [comment]
@rhubarb-geek-nz,

> If it was defined as an object and Write-Debug used ToString then you could have delayed the expansion by creating an object where the message is expanded and formatted in the ToString method.

Agree, it would be nice if the concerned commands would have something like a `[ScriptBlock]$MessageScript` (mutual exclusive with `-Message`) parameter.

Wishful thinking:
```PowerShell
Write-Debug { "This takes 3 seconds $(Sleep 3)" }
```

**Comment by jhoneill** [comment]
Side track... I have a write-logfile function, and in one of my scripts that call scripts I have proxy functions for Write-Information, Write-Debug, Write-Error, write-warning, and Write-verbose so they all go to a big text file and then call the underlying write- command so the info is still visible. It wouldn't take much to have a Write-Debug proxy function which did processed script blocks, but only if the debugpreference indicated they would be output.

**Comment by iRon7** [comment]
Another side effect for this issue is that a (debug) statement as:

```PowerShell
Write-Debug "Password: $(ConvertFrom-SecureString -AsPlainText $SecureString)"
```

will (presumably) reveal the password in memory even it is not displayed (for debugging).
